### PR TITLE
Refactor baz.thrift and TChannel runtime

### PR DESCRIPTION
**Pull request by ChuntaoLu**
This PR refactors baz.thrift to cover all the req/resp combinations, and moves the thrift method dispatch logic from gencode to runtime.

This PR is stacked on top of https://github.com/uber/zanzibar/pull/56

### CODE REVIEW

File: examples/example-gateway/endpoints/baz/silly_noop.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+	"endpointType": "http",
+	"endpointId": "baz",
+	"handleId": "sillyNoop",
+	"thriftFile": "endpoints/baz/baz.thrift",
+	"thriftFileSha": "{{placeholder}}",
+	"thriftMethodName": "SimpleService::SillyNoop",
+    "workflowType": "tchannelClient",
```
Comment (Raynos) [inline examples/example-gateway/endpoints/baz/silly_noop.json]: you keep mixing spaces vs tabs. 

Can you evaluated [eclint](https://github.com/jedmao/eclint) and add it to zanzibar for .json and .tmpl files ?

Comment (ChuntaoLu) [inline examples/example-gateway/endpoints/baz/silly_noop.json]: fixed, I will look into eclint

File: runtime/tchannel_server.go [hunk 2]
```diff
@@ -42,11 +42,11 @@ import (
 type PostResponseCB func(ctx context.Context, method string, response RWTStruct)
 
 type handler struct {
-	server         TChanServer
+	handler        TChanHandler
```
Comment (Raynos) [inline runtime/tchannel_server.go]: handler.handler is a bit weird.

File: examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go [hunk 3]
```diff
@@ -63,7 +63,7 @@ func TestCallSuccessfulRequestOKResponse(t *testing.T) {
 	}
 	defer gateway.Close()
 
-	gateway.TChannelBackends()["baz"].Register(bazServer.NewServerWithSimpleServiceCall(call))
+	gateway.TChannelBackends()["baz"].Register(bazServer.NewSimpleServiceCallHandler(call))
```
Comment (Raynos) [inline examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go]: Somehow this feels wrong.

`Register({ Service, Method, RequestHandler })` feels wrong compared to `Register("Service::Method", RequestHandler)`

It's weird that we register an interface that contains methods to get the arg1 name.

Definitely not symmetrical with registering http endpoints

File: examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go [hunk 4]
```diff
@@ -25,7 +25,7 @@ package baz
 import (
```
Comment (ChuntaoLu) [inline examples/example-gateway/endpoints/baz/baz_simpleservice_method_call_test.go]: will do

File: runtime/tchannel_server.go (hunk 2 above)
Comment (ChuntaoLu) [inline runtime/tchannel_server.go]: will rename

**Pull request merged**